### PR TITLE
#3 Facilitate executing new Alchemy installation against ganache

### PR DESCRIPTION
**Pull request by dkent600**
Implements #3, Facilitate executing new Alchemy installation against ganache.

Pairs with the [recent arc-js PR](https://github.com/daostack/arc-js/pull/23)

### CODE REVIEW

**Comment by dkent600** [comment]
@orenyodfat 

> Please add travis tests.

Adding tests is part of another PR:  https://github.com/daostack/Alchemy-Aurelia/issues/10.  If I add travis test support now it would be as an empty .travis.yml file.  Is that what you want?

> Please remove commented code.

My understanding is that your concern here is about "messy" code in NPM packages.  This repo is not being published as an npm package (what would be the point?).  While there is certainly some commented-code that could easily be removed, there are other bits that are important in the flow of ongoing work and I would really prefer not to remove them.

**Comment by orenyodfat** [comment]
@dkent600 

> Adding tests is part of another PR: #10. If I add travis test support now it would be as an empty .travis.yml file. Is that what you want?

It is very handful and useful to add an automated tests which try to cover code you submit.
For the one who submit the pr and the one who reviewing it.

> My understanding is that your concern here is about "messy" code in NPM packages. This repo is not being published as an npm package (what would be the point?). While there is certainly some commented-code that could easily be removed, there are other bits that are important in the flow of ongoing work and I would really prefer not to remove them. Is that OK with you?

This repo is being publish as part of DAOstack github.It does not matter if it produced NPM package or not. If you must have commented code so keep it in a branch on a forked repo . 
I would suggest to remove any commented code which can be removed.

File: README.md [hunk 1]
```diff
@@ -13,39 +14,81 @@ At this time there are no web server-side components.
...
+```
 
-4. start the application:
-  ```shell
-    npm start serve
-  ```
+3. Fire up ganache (testrpc) in a separate window:
+
+```shell
+npm start arc-js.ganache.runAsync
```
Comment (orenyodfat) [inline README.md]: I would suggest to use fire up a none cli ganache in this case : https://github.com/trufflesuite/ganache
and (npm install and npm start).
For running on ganache there is no meaning to link the addresses  to https://ganache.etherscan.io/ as the code currently doing.
For the case of ganache the link should be something like : HTTP://127.0.0.1:7545/<address> so one can get a real view of the address.

Comment (dkent600) [reply README.md]: @orenyodfat Sorry, I don't understand what you're saying here:

> I would suggest to use fire up a none cli ganache in this case : 
> https://github.com/trufflesuite/ganache
>and (npm install and npm start).

Are you suggesting that `npm start` with no arguments fire up ganache (`arc-js.ganache.runAsync`) and migrate the contracts into it (`arc-js.migratContracts`)?  (I think that is doable)

Comment (dkent600) [reply README.md]: @orenyodfat Great suggestion:

> For running on ganache there is no meaning to link the addresses to https://ganache.etherscan.io/ as the code currently doing.
> For the case of ganache the link should be something like : HTTP://127.0.0.1:7545/

Comment (dkent600) [reply README.md]: @orenyodfat 

> It is very handful and useful to add an automated tests which try to cover code you submit.
> For the one who submit the pr and the one who reviewing it.
>It would also catch simple issues like failing at "npm install"

At minimum, I can create a travis script that will test the scripts I've added in this PR by running ganache and doing a migration, the same thing that the arc-js travis test script does.  That will be a start on fleshing out more coding tests later one.

Comment (dkent600) [reply README.md]: @orenyodfat 

> For running on ganache there is no meaning to link the addresses to https://ganache.etherscan.io/ as the code currently doing.
> For the case of ganache the link should be something like : HTTP://127.0.0.1:7545/

I am not able to find any way to explore transactions on ganache that is out-of-the-box.

Comment (orenyodfat) [reply README.md]: @dkent600 
"I am not able to find any way to explore transactions on ganache that is out-of-the-box."
try to use ganache none cli .

Comment (dkent600) [reply README.md]: @orenyodfat 

OK, you mean _non_ cli ganache as in a GUI explorer for ganache.  I didn't know they had such an app.

You are suggesting that the ganache explorer be:

1) something that can display information about a transaction via the url at port 7545
2) launched by the script `arc-js.ganache.runAsync`

It is not trivial, at the least, to install and build this application for the user.  I think the suggestion is out-of-scope for this PR.  The cli ganache works.

So for the transaction links that are currently broken with ganache, I can change them to create the non-cli app url.  The url will not work unless the user happens to have the explorer running.  Or I can just make the urls not hot if the demo is targetting ganache.

Which option do you prefer?

Comment (orenyodfat) [reply README.md]: > So for the transaction links that are currently broken with ganache, I can change them to  create the non-cli app url. The url will not work unless the user happens to have the explorer running. Or I can just make the urls not hot if the demo is targetting ganache.
Which option do you prefer? 

I prefere to make this links meaningful (hot).
Not sure why you cannot fire up the none-cli ganache app...(browser based) and migrate to that?

Comment (dkent600) [reply README.md]: @orenyodfat 

> Not sure why you cannot fire up the none-cli ganache app...(browser based) and migrate to that?

It is because the application has to be installed, just like with parity, and that is not a simple matter of running npm.

Changing the url is an easy and low-risk change, I'll go ahead and roll it into this PR.

Comment (orenyodfat) [reply README.md]: >It is because the application has to be installed, just like with parity, and that is not a simple matter of running npm.

it can be fired up by npm install & npm start . (this is how I am bringing that up)

Comment (dkent600) [reply README.md]: You want `npm install` in what context?

**Comment by dkent600** [comment]
@orenyodfat Regarding adding travis tests:  I have introduced a .travis.yml file, but travis won't start using it until it appears in the master branch, ie, until this PR is merged.

File: README.md (hunk 1 above)
Comment (dkent600) [reply README.md]: @orenyodfat 

> I prefer to make this links meaningful (hot).

If the ganache GUI client supports automation via URL requests like this, I have not found a way to get it to work, neither by experimentation nor documentation.  Any suggestions?

**Comment by dkent600** [comment]
@orenyodfat I just committed changes that removed a lot of commented-out code and froze all the address links when running against ganache.  I previously introduced a .travis.yml file.  So I believe I've addressed all your concerns (unless you know of some way to implement your request to make the links work against the ganache GUI application).

**Comment by dkent600** [comment]
> Regarding adding travis tests: I have introduced a .travis.yml file, but travis won't start using it until it appears in the master branch, ie, until this PR is merged.

Actually I was wrong, it has begun running the test.

**Comment by orenyodfat** [comment]
@dkent600 please help travis to become green .

**Comment by dkent600** [comment]
[Approve and merge the outstanding end-to-end PR so we can submit a PR that fixes the GlobalConstraints dashboard.](https://trello.com/c/N84vtb98/38-approve-and-merge-the-outstanding-end-to-end-pr-so-we-can-submit-a-pr-that-fixes-the-globalconstraints-dashboard)

**Comment by leviadam** [comment]
Hi @dkent600, do we need to commit this file `.vscode/launch.json`?
Isn't it a workspace file?

File: README.md [hunk 2]
```diff
@@ -1,10 +1,11 @@
-# DAOStack Alchemy - Creating DAOs for an Emergent Future
+# DAOstack Alchemy - Creating DAOs for an Emergent Future
```
Comment (leviadam) [inline README.md]: Not sure about the word emergent here

**Comment by leviadam** [comment]
@dkent600 this branch is now not working for me. Namely, I get errors when trying to start the demo. Please check.

**Comment by dkent600** [comment]
@leviadam 

> this branch is now not working for me.

Can you please give more specifics?  What do you mean "when trying to start the demo" and what are the error messages, where are you seeing them?

Thanks

File: README.md (hunk 2 above)
Comment (dkent600) [reply README.md]: I like it because it reflects a fact in a new and useful way of thinking that is very much part of the epistemology that has led to the existence of DAOstack.  And it is prosaic.

Comment (leviadam) [reply README.md]: I think it is better we avoid it for now.

**Comment by leviadam** [comment]
I removed the repo completely. Cloned it again. Did `npm install`. Then `npm start`, I get:
```
npm start  

> daostack-alchemy@0.0.0-alpha.3 start /home/adam/cryptoDev/Ethereum/daostack/Alchemy-Aurelia
> nps

nps is executing `default` : nps arc-js.ganache.runAsync && nps arc-js.migrateContracts && nps build.development && nps browse
nps is executing `arc-js.ganache.runAsync` : npm explore daostack-arc-js -- npm start test.ganache.runAsync

> daostack-arc-js@0.0.0-alpha.14 start /home/adam/cryptoDev/Ethereum/daostack/Alchemy-Aurelia/node_modules/daostack-arc-js
> nps "test.ganache.runAsync"

nps is executing `test.ganache.runAsync` : osascript -e 'tell application "Terminal"' -e 'tell application "System Events" to keystroke "t" using {command down}' -e 'do script "cd /home/adam/cryptoDev/Ethereum/daostack/Alchemy-Aurelia/node_modules/daostack-arc-js && npm start test.ganache.run" in front window' -e 'end tell'
/bin/sh: 1: osascript: not found
The script called "test.ganache.runAsync" which runs "osascript -e 'tell application "Terminal"' -e 'tell application "System Events" to keystroke "t" using {command down}' -e 'do script "cd /home/adam/cryptoDev/Ethereum/daostack/Alchemy-Aurelia/node_modules/daostack-arc-js && npm start test.ganache.run" in front window' -e 'end tell'" failed with exit code 127 https://github.com/kentcdodds/nps/blob/v5.7.1/other/ERRORS_AND_WARNINGS.md#failed-with-exit-code
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! daostack-arc-js@0.0.0-alpha.14 start: `nps "test.ganache.runAsync"`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the daostack-arc-js@0.0.0-alpha.14 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/adam/.npm/_logs/2018-01-24T14_10_03_400Z-debug.log
npm ERR! weird error 1
The script called "arc-js.ganache.runAsync" which runs "npm explore daostack-arc-js -- npm start test.ganache.runAsync" failed with exit code 1 https://github.com/kentcdodds/nps/blob/v5.7.1/other/ERRORS_AND_WARNINGS.md#failed-with-exit-code
The script called "default" which runs "nps arc-js.ganache.runAsync && nps arc-js.migrateContracts && nps build.development && nps browse" failed with exit code 1 https://github.com/kentcdodds/nps/blob/v5.7.1/other/ERRORS_AND_WARNINGS.md#failed-with-exit-code
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! daostack-alchemy@0.0.0-alpha.3 start: `nps`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the daostack-alchemy@0.0.0-alpha.3 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/adam/.npm/_logs/2018-01-24T14_10_03_432Z-debug.log
```

**Comment by dkent600** [comment]
@leviadam Per the Readme:

> If `npm start` doesn't work for you, and for more information about the scripts involved, see the following section "Build and Run".

Under there you will find:

### Fire up ganache (testrpc) in a separate window

```shell
npm start arc-js.ganache.runAsync
```

Note: If the window didn't fire up in your OS, then run this in a separate window of your own creation:

```shell
npm start arc-js.ganache.run
```

File: package.json [hunk 3]
```diff
@@ -27,8 +27,13 @@
       "ts"
     ],
     "transform": {
-      "^.+\\.(ts|tsx)$": "<rootDir>/node_modules/ts-jest/preprocessor.js"
+      "^.+\\.(ts|tsx)$": "<rootDir>/node_modules/ts-jest/preprocessor.js",
```
Comment (orenyodfat) [inline package.json]: > "^.+\\.(ts|tsx)$": "<rootDir>/node_modules/ts-jest/preprocessor.js",

I guess it will be nice to add a comment re this wird line

File: README.md [hunk 4]
```diff
@@ -13,38 +14,95 @@ At this time there are no web server-side components.
...
-4. start the application:
-  ```shell
-    npm start serve
-  ```
+```shell
+npm install
+npm start
+```
+
+If `npm start` doesn't work for you, and for more information about the scripts involved, see the following section "Build and Run".
```
Comment (orenyodfat) [inline README.md]: > If `npm start` doesn't work for you 

Why ?  Is there a known cases where it will not run ?
If yes, can you list that ?
If not i guess this line should be updated :  

> For more information about the scripts involved, see the following section "Build and Run".

File: package.json (hunk 3 above)
Comment (dkent600) [reply package.json]: Can't comment .json files.

**Comment by dkent600** [comment]
@leviadam 

> do we need to commit this file .vscode/launch.json?

I have removed this folder.

File: README.md (hunk 4 above)
Comment (dkent600) [reply README.md]: I honestly am not sure which OSs will work and which won't, but it definitely won't work for all of them, thus it is anticipated that `npm start` won't work for everyone, thus I prefer to be clear about that in the comment.

If we want we can create a git issue to try to make runAsync on every OS

File: package.json (hunk 3 above)
Comment (dkent600) [reply package.json]: I would note that this line is no less weird than many others in that file and especially in webpack.config.js.

**Comment by leviadam** [comment]
@dkent600 I checked also the other script, not just `npm start`, they also do not work.

**Comment by dkent600** [comment]
@leviadam 

Are you saying that `npm start arc-js.ganache.run` did not work for you?

Need more details as before...thanks

**Comment by dkent600** [comment]
@leviadam I removed the word "Emergent"

File: README.md (hunk 4 above)
Comment (leviadam) [reply README.md]: But we do have a way to run async on all OS, no?
We do it in daostack in the `test` script.
It works quite well AFAIK.

Comment (dkent600) [reply README.md]: > But we do have a way to run async on all OS, no?

I could probably put something together, but would prefer to make that a separate ticket.  The instructions are there for the case when the Async script doesn't work.

**Comment by dkent600** [comment]
@leviadam With regard to your problems running the scripts on linux:

I ran them in a Bash shell.  This is not a full installation of Ubuntu, but is likely close enough.

I did encounter some issues relating to npm packages and one issue of a case-sensitive filename mismatch.  Not knowing exactly what your issues were, I can't know that these are the same issues you encountered.

In any case, I made changes and added a couple trouble-shooting instructions in the readme.

Can you please try again?  If you encounter problems, please check the readme for new suggestions that might help.

Thanks

**Comment by leviadam** [comment]
Still not working.

**Comment by dkent600** [comment]
Created #24 to make the `arc-js.ganache.runAsync` work with Linux

**Comment by dkent600** [comment]
For some reason this PR isn't showing it in the GUI, but the tests are succeeding:  

https://travis-ci.org/daostack/Vanilla

File: README.md [hunk 5]
```diff
@@ -1,10 +1,11 @@
-# DAOStack Alchemy - Creating DAOs for an Emergent Future
+# DAOstack Alchemy - Creating DAOs for a Collective Intelligence
```
Comment (orenyodfat) [inline README.md]: "a Collective Intelligence" :)

File: README.md [hunk 6]
```diff
@@ -1,10 +1,11 @@
...
 
 ## What is Alchemy?
 
-DAOStack Alchemy is a browser application that enables anyone with an account on the Ethereum blockchain, and STK tokens to spend, to easily create and work with DAOs on Ethereum that run on predefined, configurable smart contracts in a network of organizations, collaborators and investors that collaborate on the blockchain.
+DAOstack Alchemy is a browser application that enables anyone with an account on the Ethereum blockchain, and STK tokens to spend, to easily create and work with DAOs on Ethereum that run on predefined, configurable smart contracts in a network of organizations, collaborators and investors that collaborate on the blockchain.
```
Comment (orenyodfat) [inline README.md]: > DAOstack Alchemy is a browser application that enables anyone with an account on the Ethereum blockchain, and STK tokens to spend, to easily create and work with DAOs on Ethereum that run on predefined, configurable smart contracts in a network of organizations, collaborators and investors that collaborate on the blockchain.

I will revised this long sentence. It is long and not so accurate.
Is that a Vanilla or Alchemy repo ?
STK - > GEN

File: README.md [hunk 7]
```diff
@@ -1,10 +1,11 @@
...
 
 ## Implementation
+
 The application uses the standards-compliant and highly modular browser-side framework [Aurelia](http://aurelia.io) (hence "Aurelia" in the name of the repository).
```
Comment (orenyodfat) [inline README.md]: Aurelia is not the name of the repo.

File: README.md [hunk 8]
```diff
@@ -1,10 +1,11 @@
...
 
 Browser-side the application uses TypeScript/ECMAScript 2016+.
```
Comment (orenyodfat) [inline README.md]: > n uses TypeScript/ECMAScript 2016+. 

Is that significant ?

File: README.md [hunk 9]
```diff
@@ -13,38 +14,117 @@ At this time there are no web server-side components.
 
 ## DaoStack-Arc-Js
 
-Alchemy uses a library of reusable and configurable Ethereum smart contracts called "DaoStack-Arc-Js".  You can see all of the source code for those contracts and even contribute to the project [here](https://github.com/daostack/arc-js).
+Alchemy uses a library of reusable and configurable Ethereum smart contracts called "DaoStack-Arc-Js". You can see all of the source code for those contracts and even contribute to the project [here](https://github.com/daostack/arc-js).
```
Comment (orenyodfat) [inline README.md]: > [here](https://github.com/daostack/arc-js).

This is not the contracts repo.

File: src/app.ts [hunk 10]
```diff
@@ -11,100 +11,98 @@ export class App {
...
+    (<any>$('body')).bootstrapMaterialDesign({ global: { label: { className: "bmd-label-floating" } } });
   }
-    
+
   configureRouter(config: RouterConfiguration, router: Router) {
 
-    config.title = 'DAOStack Alchemy';
-    // config.options.pushState = true;
-    // config.options.root = '/';
+    config.title = 'DAOstack Alchemy';
```
Comment (orenyodfat) [inline src/app.ts]: Vanille

File: README.md (hunk 6 above)
Comment (dkent600) [reply README.md]: Good catch on GEN and Vanille.  As for the run-on sentence, I'm having a hard time making it shorter.  But I edited it a little to make it flow a little better.  It works for me, but maybe I'm more accustomed to reading long sentences...

File: README.md (hunk 7 above)
Comment (dkent600) [reply README.md]: Good catch, done.

File: README.md (hunk 8 above)
Comment (dkent600) [reply README.md]: To coders, yes.

File: README.md (hunk 9 above)
Comment (dkent600) [reply README.md]: Some of the references to arc-js will have to wait until I upgrade Vanille to work with a newer (and refactored) package of arc.js.  That needs to be in another PR.  (see https://github.com/daostack/Vanille/issues/25)

Other references to "arc-js" will have to remain in any case, namely in the names of the scripts -- the dot is a special character in those names.

File: src/app.ts (hunk 10 above)
Comment (dkent600) [reply src/app.ts]: Done

File: README.md (hunk 9 above)
Comment (orenyodfat) [reply README.md]: You can see all of the source code for those contracts and even contribute to the project [here](https://github.com/daostack/arc-js).

Contracts are in Arc (not arc-js).

File: README.md (hunk 8 above)
Comment (orenyodfat) [reply README.md]: I am not a coder..what is it ?

File: package.json [hunk 11]
```diff
@@ -1,16 +1,16 @@
...
+  "description": "Creating DAOs for a Collective Intelligence",
   "main": "dist/app.bundle.js",
   "repository": {
-    "url": "git+ssh://git@github.com/daostack/Alchemy-Aurelia.git",
+    "url": "git+ssh://git@github.com/daostack/Vanille.git",
     "type": "git"
   },
-  "author": "DAOStack",
+  "author": "DAOstack",
   "license": "CC0-1.0",
```
Comment (orenyodfat) [inline package.json]: > "license": "CC0-1.0", 

I though we are under GPL .

Comment (dkent600) [reply package.json]: Yes, that change is sitting in another branch waiting for this PR to be approved.  Can we just let it be there?

Comment (orenyodfat) [reply package.json]: Please open an issue for that so it could be tracked.

or

Change that to

>"license": "GPL-3.0", 

-copy the license file from https://github.com/daostack/Arc/blob/master/LICENSE

File: README.md (hunk 8 above)
Comment (dkent600) [reply README.md]: Sorry, poor wording, I think you *are*  a coder, and a good one  :-)

What I meant was that it is important to website developers who might have an interest in coding the application, or who might just be curious what languages were used.  TypeScript is a  language, ECMAScript 2016 is a version of JavaScript.

However, I think to be safe, I think it should be 2015 (ES6), not 2016 (ES7), so I have changed it.

File: README.md (hunk 9 above)
Comment (dkent600) [reply README.md]: OK, see what you think...

**Comment by dkent600** [comment]
@orenyodfat If you have any further comments about documentation that don't pertain to running scripts, can you please just add them to this issue:  #27, not to this PR?

Thanks!
